src/markdown-tests/build.ts

```diff
@@ -0,0 +1,217 @@
+  const result: Describe = {
+    name: "root",
+    children: [],
+    parent: null,
+    level: 0,
+  };
+  let currentDescribe = result;
+  let lastInput: CodeBlock | null = null;
+  for (const tag of extractTags(markdown)) {
+    if (tag.kind === "Header") {
+      if (tag.level > currentDescribe.level + 1) {
+        throw new Error("Header level must increase by at most one.");
+      }
+      while (tag.level <= currentDescribe.level) {
+        currentDescribe = currentDescribe.parent!;
+      }
+      const newDescribe = {
+        name: tag.title,
+        children: [],
+        parent: currentDescribe,
+        level: tag.level,
+      };
+      currentDescribe.children.push(newDescribe);
+      currentDescribe = newDescribe;
+    } else if (!tag.args.includes("skip")) {
+      if (tag.language === "polygolf" && tag.args.length === 0) {
+        lastInput = tag;
+      } else {
+        if (lastInput === null) {
+          throw new Error("No input block found!");
+        }
+        const lang = findLang(tag.language);
+        if (lang === undefined) {
+          throw new Error(`Unknown language ${tag.language}`);
+        }
+        currentDescribe.children.push({
+          input: lastInput.content,
+          language: lang.name,
+          commands: tag.args,
+          output: tag.content,
+        });
+      }
+    }
+  }
+  function removeEmptyDescribes(describe: Describe) {
+    for (const child of describe.children) {
+      if ("children" in child) {
+        removeEmptyDescribes(child);
+      }
+    }
+    describe.children = describe.children.filter(
+      (x) => "input" in x || x.children.length > 0
+    );
+  }
+  removeEmptyDescribes(result);
+  if (result.children.length === 1 && "children" in result.children[0])
+    return result.children[0];
+  return result;
+}
+
+/** Extracts relevant markdown sections:
+ *  Headlines and codeblocks.
+ */
+function extractTags(markdown: string): (Header | CodeBlock)[] {
+  const result: (Header | CodeBlock)[] = [];
+  const regex = /\n(#+)([^\n]+)\n|```([^`\n]*)\n([^`]*)\n```/gs;
+  let m;
+  while ((m = regex.exec("\n" + markdown)) !== null) {
+    if (m[1] !== undefined) {
+      result.push({
+        kind: "Header",
+        level: m[1].length,
+        title: m[2].trim(),
+      });
+    } else {
+      const title = m[3].trim().split(" ");
+      if (m[3].trim().length === 0) {
+        throw new Error(
+          "Codeblocks are required to have language annotation. If a codeblock should not be included in testing, use `skip` command: `python skip`."
+        );
+      }
+      result.push({
+        kind: "CodeBlock",
+        language: title[0],
+        args: title.slice(1),
+        content: m[4].trim(),
+      });
+    }
+  }
+  return result;
+}
+
+/** Emits a single markdown test. */
+function emitSuite(describe: Describe): string {
+  return `import parse from "frontend/parse";
+import applyLanguage from "@/common/applyLanguage";
```

Comment:
Might want a comment here to the effect of: the `@/` path is defined relatively in tsconfig.json, pointing to the `src` directory

Reply:
Is this not a common thing to have?

Reply:
I spent some time googling TypeScript `@/`, thinking it was a language feature. Would help clarity for future readers